service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/CreateAzureVmStep.java

```diff
@@ -104,19 +110,12 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
                   azureCloudContext.getAzureResourceGroupId(), networkResource.getNetworkName());
 
       var createNic =
-          computeManager
-              .networkManager()
-              .networkInterfaces()
-              .define(String.format("nic-%s", resource.getVmName()))
-              .withRegion(resource.getRegion())
-              .withExistingResourceGroup(azureCloudContext.getAzureResourceGroupId())
-              .withExistingPrimaryNetwork(existingNetwork)
-              .withSubnet(networkResource.getSubnetName())
-              .withPrimaryPrivateIPAddressDynamic()
-              .withExistingPrimaryPublicIPAddress(
-                  existingAzureIp) // TODO this needs to be updated to support not exposing public
-              // IP
-              .create();
+          createNetworkInterface(
```

Comment:
I think this step has idempotency problems. If this step is retried or there is a pod failover, then the newly created NIC will be orphaned, right?

I think you need a sequence something like this:
- step 1: Compute and store the NIC name in the working map (Or indicate "use existing ip")
- step 2:
  - do: if NIC exists, return success. Otherwise, create the NIC. 
  - undo: if NIC exists, delete it. Return success
- step 3: the CreateAzureVmStep
  - do: compute the inputs (nothing is created), then create the VM
  - undo: delete the VM if it exists

You could probably squeeze step 2 into step 3, but I think more, simpler steps make reasoning about failure and idempotency easier.

Reply:
After a side conversation with @sergiygetlin, I understand that this is the same issue that is already captured in the ticket TOAZ-114. I would like that work prioritized so we are not accumulating tech debt in the WSM code.